main/src/library/MutList.flix

```diff
@@ -18,145 +18,404 @@
+            foldLeftHelper(f, s1, v, len, i + 1)
+        else
+            s
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `a` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, a[0]), a[1])..., xn)`.
+    ///
+    pub def foldRight(f: (a, b) -> b & e, s: b, v: MutList[a]): b & Impure =
+        let len = length(v);
+        foldRightHelper(f, s, v, len - 1)
+
+    ///
+    /// Helper function for `foldRight`.
+    ///
+    def foldRightHelper(f: (a, b) -> b & e, s: b, v: MutList[a], i: Int32): b & Impure =
+        if (i < 0)
+            s
+        else
+            let s1 = f(get(i, v), s) as & Impure;
+            foldRightHelper(f, s1, v, i - 1)
+
+    ///
+    /// Alias for `reduceLeft`.
+    ///
+    pub def reduce(f: (a, a) -> a & e, v: MutList[a]): Option[a] & Impure = reduceLeft(f, v)
+
+    ///
+    /// Applies `f` to all elements in `v` going from left to right until a single value `v` is obtained. Returns `Some(v)`.
+    ///
+    /// Returns `None` if `v` is empty.
+    ///
+    pub def reduceLeft(f: (a, a) -> a & e, v: MutList[a]): Option[a] & Impure = foldLeft((acc, x) -> match acc {
+        case Some(y) => Some(f(y, x))
+        case None => Some(x)
+    }, None, v)
+
+    ///
+    /// Applies `f` to all elements in `v` going from right to left until a single value `v` is obtained. Returns `Some(v)`.
+    ///
+    /// Returns `None` if `v` is empty.
+    ///
+    pub def reduceRight(f: (a, a) -> a & e, v: MutList[a]): Option[a] & Impure = foldRight((x, acc) -> match acc {
+        case Some(y) => Some(f(x, y))
+        case None => Some(x)
+    }, None, v)
+
+    ///
+    /// Removes all elements from the given mutable list `v`.
+    ///
+    pub def clear!(v: mut MutList[a]): Unit & Impure =
+        let MutList(a, l) = v;
+        a := Array.new(default, Array.length(deref a));
+        l := 0
 
     ///
     /// Returns a shallow copy of the given mutable list `v`.
     ///
-    pub def copy(v: MutList[a]): MutList[a] = ???
+    pub def copy(v: MutList[a]): MutList[a] & Impure =
+        let MutList(a, l) = v;
+        let len = deref l;
+        MutList(ref Array.slice(deref a, 0, Array.length(deref a)), ref len)
 
     ///
     /// Optionally removes and returns the last element in the given mutable list `v`.
     ///
-    pub def pop(v: mut MutList[a]): Option[a] = ???
+    pub def pop!(v: mut MutList[a]): Option[a] & Impure =
+        let MutList(a, l) = v;
+        let len = deref l;
+        if (len > 0)
+            let last = (deref a)[len - 1];
+            l := len - 1;
+            Array.put(deref a, len - 1, default);
+            compress(v);
+            Some(last)
+        else
+            None
 
     ///
     /// Inserts the given element `x` at the end of the given mutable list `v`.
     ///
-    pub def push(v: mut MutList[a], x: a): Unit = ???
+    pub def push!(x: a, v: mut MutList[a]): Unit & Impure = insert(x, length(v), v)
```

Comment:
I would rather that this had its own specialized implementation.

Reply:
Also, when you run out of space, this should double the size of the backing array. Similar to compress :)